mantidproject/mantid
# Workbench: Update plots if the workspace data changes

**Description of work.**

Adds an ADS observer to the FigureManager in the workbench. The line and color plots now respond appropriately when workspaces are replaced or deleted.

**Report to:** [nobody]. 

**To test:**

Data: 
  - `ISIS_Powder/input/HRPD66063_focused.nxs` in the `Testing/Data/SystemTest` directory
  - `tests/analysis/reference/MARIReductionSum.nxs` in the `Testing/SystemTest` directory

Use the following script to produce some test workspaces:

```python
focussed = Load('ExternalData/Testing/Data/SystemTest/ISIS_Powder/input/HRPD66063_focused.nxs')
non_common_bins = CloneWorkspace('66063.raw-ResultD-1')
ConjoinWorkspaces(non_common_bins, '66063.raw-ResultD-2')
ConjoinWorkspaces(non_common_bins, '66063.raw-ResultD-3')

common_bins = Load('ExternalData/Testing/SystemTests/tests/analysis/reference/MARIReductionSum.nxs')
```

Use the produced workspaces to test various plots still work:
* use the items in `focussed` to test line plots
* use the both `non_common_bins` and `common_bins` to test colorfill plots

When each of the plots of displayed try deleting workspaces and ensure that the relevant plots are closed when the linked workspace is deleted. For the line plots add multiple curves from different workspaces to the same plot and check that deleting a single workspace only deletes the curves from that workspace.

Try replacing the workspaces using, e.g. `Rebin`, and confirm that the data on any attached plots is updated as expected. 

Fixes #24000 

*This does not require release notes* because **it is part of the unreleased workbench**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
